Fix print of ValuesNode from HiveFilterPushDown

feilong-liu opened the pull request:
### What the issue is?
In HiveFilterPushdown, a table scan node will be replaced with an empty Value node if the table scan node produces empty output. It causes confusion when checking the query plan, as the table scan node is not present.

### What the fix is?
In the ValuesNode, add one optional property "origTableName", which is the table name scanned converted to the ValuesNode. It defaults to empty for other cases. In the printer classes, it will output this information if  "origTableName" is not empty.

### Test plan
Test locally with a simple scan query, manually trigger the HiveFilterPushDown optimization and check the explain output.

**Current behavior:**
presto:tpch> explain select * from nation where nationkey=100;
                                                          Query Plan                                                           
 - Output[nationkey, name, regionkey, comment] => [nationkey:bigint, name:varchar(25), regionkey:bigint, comment:varchar(152)] 
         Estimates: {rows: 0 (0B), cpu: 0.00, memory: 0.00, network: 0.00}                                                     
     - Values => [nationkey:bigint, name:varchar(25), regionkey:bigint, comment:varchar(152)]                                  
             Estimates: {rows: 0 (0B), cpu: 0.00, memory: 0.00, network: 0.00}                                                 
                                                                                                                               
(1 row)

**With this diff:**
presto:tpch> explain select * from nation where nationkey=100;
                                                                                                                                                               Query Plan                                                                                                                                                            
 - Output[nationkey, name, regionkey, comment] => [nationkey:bigint, name:varchar(25), regionkey:bigint, comment:varchar(152)]                                                                                                                                                                                                       
         Estimates: {rows: 0 (0B), cpu: 0.00, memory: 0.00, network: 0.00}                                                                                                                                                                                                                                                           
     - Values converted from TableScan[TableHandle {connectorId='hive', connectorHandle='HiveTableHandle{schemaName=tpch, tableName=nation, analyzePartitionValues=Optional.empty}', layout='Optional[tpch.nation{domains={nationkey=[ [["100"]] ]}}]'}] => [nationkey:bigint, name:varchar(25), regionkey:bigint, comment:varchar(152)]
             Estimates: {rows: 0 (0B), cpu: 0.00, memory: 0.00, network: 0.00}                                                                                                                                                                                                                                                       



(1 row)


```
== NO RELEASE NOTE ==
```

kaikalur commented on presto-spi/src/main/java/com/facebook/presto/spi/plan/ValuesNode.java:
```diff
@@ -37,13 +37,15 @@
 {
     private final List<VariableReferenceExpression> outputVariables;
     private final List<List<RowExpression>> rows;
+    private final Optional<String> origTableName;
```
Just call it label or something and also maybe use the alias that the use gives to values nodes as the label for other situations.

kaikalur replied:
Can you also give a warning (if you are not already) when a table is eliminated?

feilong-liu replied:
> Can you also give a warning (if you are not already) when a table is eliminated?

Added warning in HiveFilterPushDown, it will output warnings like

> WARNING: Table 'TableHandle {connectorId='hive', connectorHandle='HiveTableHandle{schemaName=tpch, tableName=nation, analyzePartitionValues=Optional.empty}', layout='Optional[tpch.nation{domains={nationkey=[ [["100"]] ]}}]'}' is converted to ValuesNode by HiveFilterPushdown

> Just call it label 

Rename it to valuesNodeLabel

> also maybe use the alias that the use gives to values nodes as the label for other situations.

Looking into it now.

kaikalur replied:
> Added warning in HiveFilterPushDown, it will output warnings like
> 
> > WARNING: Table 'TableHandle {connectorId='hive', connectorHandle='HiveTableHandle{schemaName=tpch, tableName=nation, analyzePartitionValues=Optional.empty}', layout='Optional[tpch.nation{domains={nationkey=[ [["100"]] ]}}]'}' is converted to ValuesNode by HiveFilterPushdown
> 

ValueNode is a plan concept. So make it clear to the user what it did exactly. I think we use it only the partition is empty. So make that explicit and say:

table ... returned 0 rows

or something like that

feilong-liu replied:
> ValueNode is a plan concept. So make it clear to the user what it did exactly. I think we use it only the partition is empty. So make that explicit and say:
>
> table ... returned 0 rows

Changed the warning to
> WARNING: Table 'TableHandle {connectorId='hive', connectorHandle='HiveTableHandle{schemaName=tpch, tableName=nation, analyzePartitionValues=Optional.empty}', layout='Optional[tpch.nation{domains={nationkey=[ [["100"]] ]}}]'}' **returns 0 rows, and is converted to an empty ValuesNode** by HiveFilterPushdown

> also maybe use the alias that the use gives to values nodes as the label for other situations.

I checked the code, the alias information is not directly available when creating the ValuesNode, and needs non trival changes to the call stack to pass this information. Consider that this is a good to have change and not directly related to the problem fixed in the pull request. Could we leave it as it is for now? thanks!

kaikalur replied:
Looks good. Please add a separate task for propagating the values node alias to plan

kaikalur reviewed (changes requested):
Will do more reviews later.

jainxrohit commented:
Wow that touched a lot of files.

jainxrohit commented on presto-spi/src/main/java/com/facebook/presto/spi/plan/ValuesNode.java:
```diff
@@ -37,13 +37,15 @@
+    private final Optional<String> valuesNodeLabel;
 
     @JsonCreator
     public ValuesNode(
```
Can we create a new constructor which takes the label name as empty? This will limit the number of overall changes.

kaikalur replied:
> Can we create a new constructor which takes the label name as empty? This will limit the number of overall changes.

There could be issues with the json constructor. I think it expects only one. I had to go through hoops to make it work last time when I introduced the source location field in plan node.

kaikalur commented:
Please fix the tests

feilong-liu commented:
> Please fix the tests

@kaikalur Tests fixed.

kaikalur reviewed (approved):
LGTM

jainxrohit commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:
```diff
@@ -809,7 +809,13 @@ public Void visitTableScan(TableScanNode node, Void context)
-            NodeRepresentation nodeOutput = addNode(node, "Values");
+            NodeRepresentation nodeOutput;
+            if (node.getValuesNodeLabel().isPresent()) {
+                nodeOutput = addNode(node, format("Values converted from TableScan[%s]", node.getValuesNodeLabel().get()));
```
Should we just say "Values from TableScan" ?

feilong-liu replied:
I assume that we can have corner case where one table appears in multiple scans which end up with converting to ValuesNode, Maybe output more information can be helpful?

highker commented:
Please squash all commits into one (or several as long as each commit is self-contained)

highker reviewed (commented):
@feilong-liu, maybe work offline with @kaikalur and @jainxrohit on the commit stack. This patch's commits seem to be intertwined with many other trunk commits. I won't be able to pull it down and review the patch locally.

**You can request my review once we have cleaned up the commits.**


```
b5a3fe5cd1 (HEAD -> 18019) Fix print of ValueNodes
a7ea3fc92b Fix print of ValueNodes
aeabea9559 Fix print of ValueNodes
436b0b9e44 Construct TDigest Function
f54d5d3f69 Reduce calls to metastore in materialized view optimization
cc6066a373 Fix print of ValueNodes
5b6f399c36 Update hudi-testing-data.md
d4126a2d7d Improve ARRAY_FREQUENCY efficiency.
e3212ac313 Allow catalog server to be run seperately from resource manager
7b31447701 Replace Joda-Time libraries with java.time
7d4a899a97 Migrate presto_cpp circleci workflows to presto
70ec8328e2 support hive built in aggregationimplementation
245324241c suppport aggregation implementation
55c8e29ad4 simplify accumulatorfactory constructor
4fa59e8ab5 Iceberg: Add new fields in files system table
4ad0082596 Fix print of ValueNodes
5a56cde3f5 Fix print of ValueNodes
c4227eb606 Fix broadcast OOM error code
```

feilong-liu commented:
> @feilong-liu, maybe work offline with @kaikalur and @jainxrohit on the commit stack. This patch's commits seem to be intertwined with many other trunk commits. I won't be able to pull it down and review the patch locally.
> 
> **You can request my review once we have cleaned up the commits.**
> 
> ```
> b5a3fe5cd1 (HEAD -> 18019) Fix print of ValueNodes
> a7ea3fc92b Fix print of ValueNodes
> aeabea9559 Fix print of ValueNodes
> 436b0b9e44 Construct TDigest Function
> f54d5d3f69 Reduce calls to metastore in materialized view optimization
> cc6066a373 Fix print of ValueNodes
> 5b6f399c36 Update hudi-testing-data.md
> d4126a2d7d Improve ARRAY_FREQUENCY efficiency.
> e3212ac313 Allow catalog server to be run seperately from resource manager
> 7b31447701 Replace Joda-Time libraries with java.time
> 7d4a899a97 Migrate presto_cpp circleci workflows to presto
> 70ec8328e2 support hive built in aggregationimplementation
> 245324241c suppport aggregation implementation
> 55c8e29ad4 simplify accumulatorfactory constructor
> 4fa59e8ab5 Iceberg: Add new fields in files system table
> 4ad0082596 Fix print of ValueNodes
> 5a56cde3f5 Fix print of ValueNodes
> c4227eb606 Fix broadcast OOM error code
> ```

@highker I've squashed the changes into two commits. One is to output the original table information for ValuesNode, the other is to output warning.

The changes is small but a lot of files are touched due to the change of function signature. Hence I split it into two commit hoping that it can make the review easier.

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/plan/ValuesNode.java:
```diff
@@ -37,13 +37,15 @@
 {
     private final List<VariableReferenceExpression> outputVariables;
     private final List<List<RowExpression>> rows;
+    private final Optional<String> valuesNodeLabel;
```
Could you add a comment to explain like "this is to add extra information for debugging purpose blab blab blah. It does not affect the execution of the value node blah blah blah"?

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java:
```diff
@@ -376,6 +396,8 @@ public PlanNode visitFilter(FilterNode filter, Void context)
             RowExpression replacedExpression = replaceExpression(expression, symbolToColumnMapping);
             // replaceExpression() may further optimize the expression; if the resulting expression is always false, then return empty Values node
             if (FALSE_CONSTANT.equals(replacedExpression)) {
+                StringBuilder warningMessage = new StringBuilder(format("Table '%s' returns 0 rows, and is converted to an empty ValuesNode by HiveFilterPushdown", tableScan.getTable().toString()));
+                warningCollector.add(new PrestoWarning(HIVE_TABLESCAN_CONVERTED_TO_VALUESNODE, warningMessage.toString()));
```
string builder is not recommended any more. Also we can inline the two:
```java
warningCollector.add(new PrestoWarning(
        HIVE_TABLESCAN_CONVERTED_TO_VALUESNODE,
        format("Table '%s' returns 0 rows, and is converted to an empty ValuesNode by HiveFilterPushdown", tableScan.getTable().toString())));

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java:
```diff
@@ -387,6 +409,8 @@ public PlanNode visitFilter(FilterNode filter, Void context)
 
             ConnectorTableLayout layout = pushdownFilterResult.getLayout();
             if (layout.getPredicate().isNone()) {
+                StringBuilder warningMessage = new StringBuilder(format("Table '%s' returns 0 rows, and is converted to an empty ValuesNode by HiveFilterPushdown", tableScan.getTable().toString()));
+                warningCollector.add(new PrestoWarning(HIVE_TABLESCAN_CONVERTED_TO_VALUESNODE, warningMessage.toString()));
```
same as above

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java:
```diff
@@ -422,7 +446,10 @@ public PlanNode visitTableScan(TableScanNode tableScan, Void context)
                     handle.getLayout());
+            WarningCollector warningCollector = context.getWarningCollector();
             if (pushdownFilterResult.getLayout().getPredicate().isNone()) {
+                StringBuilder warningMessage = new StringBuilder(format("Table '%s' returns 0 rows, and is converted to an empty ValuesNode by HiveFilterPushdown", tableScan.getTable().toString()));
+                warningCollector.add(new PrestoWarning(HIVE_TABLESCAN_CONVERTED_TO_VALUESNODE, warningMessage.toString()));
```
same

highker reviewed (approved):
I'm 100% sure this will break https://github.com/facebookexternal/presto-facebook repo. Can you send the corresponding PR to fix that one as well. Then we can merge then together.

kaikalur commented:
> I'm 100% sure this will break https://github.com/facebookexternal/presto-facebook repo. Can you send the corresponding PR to fix that one as well. Then we can merge then together. @jainxrohit, @kaikalur, please help to advise how to have the https://github.com/facebookexternal/presto-facebook fix.

Hmm - codesearch shows no use of ValuesNode though

highker commented:
@kaikalur, not `ValuesNode` actually. It's the SPI change adding the warning collector.

kaikalur commented:
> @kaikalur, not `ValuesNode` actually. It's the SPI change adding the warning collector.

Oh ok. Got it!

kaikalur commented:
@feilong-liu do we really need the warningsCollector param? IterativeOptimizer.Context already has a warningsCollector field. Can we use that?

feilong-liu commented:
> @feilong-liu do we really need the warningsCollector param? IterativeOptimizer.Context already has a warningsCollector field. Can we use that?

The HiveFilterPushdown optimization is not an iterative optimizer, it implements the interface `ConnectorPlanOptimizer` and has no warningsCollector available.

kaikalur commented:
> > @feilong-liu do we really need the warningsCollector param? IterativeOptimizer.Context already has a warningsCollector field. Can we use that?
> 
> @kaikalur The HiveFilterPushdown optimization is not an iterative optimizer, it implements the interface `ConnectorPlanOptimizer` and has no warningsCollector available.

I see a ConnectorSession interface that has other sundry info like client tags etc. So I think you can just add a `addWarning` method to that with a default no-op implementation and implement it properly in the HiveConnectorSession to add it to the WarningsCollector.

highker commented on presto-main/src/main/java/com/facebook/presto/FullConnectorSession.java:
```diff
@@ -74,6 +78,27 @@ public FullConnectorSession(
         this.sessionPropertyManager = requireNonNull(sessionPropertyManager, "sessionPropertyManager is null");
         this.sqlFunctionProperties = session.getSqlFunctionProperties();
         this.sessionFunctions = ImmutableMap.copyOf(session.getSessionFunctions());
+        this.warningCollector = Optional.empty();
```
We probably don't want to put warningCollector half way through the session. The session should be built with warningCollector from step 0. The place to put warningCollector into session is in `DispatchManager` where session was first created:
```java
session = sessionSupplier.createSession(queryId, sessionContext);
```

Note that `warningCollector` was created right after it because it depends on session. So in this case, you might need to change `createSession(queryId, sessionContext)` to be `createSession(queryId, sessionContext, warningCollectorFactory)` so that we create both session and warningCollector in one place.

Also, the next line `preparedQuery = queryPreparer.prepareQuery(session, query, warningCollector);` can also be simplified with `preparedQuery = queryPreparer.prepareQuery(session, query);` because warningCollector is now available in session already.

feilong-liu replied:
> The session should be built with warningCollector from step 0.

@highker I checked the code and not sure how to deal with the warningCollector when converting Session to SessionRepresentation.
I see that SessionRepresentation is a thrift struct, is it possible to put the warningCollector to the SessionRepresentation too, or it doesn't matter and we do not need to pass warningCollector into SessionRepresentation?

highker replied:
You can just use NOOP for session representation. Session representation is used for communication. We don't need warning collectors on the worker side. Make sure you have a comment on "we use NOOP to create a session from the representation as worker does not require warning collectors".

feilong-liu replied:
> you might need to change createSession(queryId, sessionContext) to be createSession(queryId, sessionContext, warningCollectorFactory) so that we create both session and warningCollector in one place.

Done

> Also, the next line preparedQuery = queryPreparer.prepareQuery(session, query, warningCollector); can also be simplified with preparedQuery = queryPreparer.prepareQuery(session, query); because warningCollector is now available in session already.

I leave it as it is for now. As changing this function will lead to changes in  a bunch of callsites of this function. Since we are to add a separate PR to clean up the code with respect to warningCollector, I think it maybe better to leave it to the following PR.

highker commented on presto-main/src/main/java/com/facebook/presto/Session.java:
```diff
@@ -492,6 +493,20 @@ public ConnectorSession toConnectorSession(ConnectorId connectorId)
                 sessionPropertyManager);
     }
 
+    public ConnectorSession toConnectorSession(ConnectorId connectorId, WarningCollector warningCollector)
+    {
+        requireNonNull(connectorId, "connectorId is null");
+
+        return new FullConnectorSession(
+                this,
+                identity.toConnectorIdentity(connectorId.getCatalogName()),
+                connectorProperties.getOrDefault(connectorId, ImmutableMap.of()),
+                connectorId,
+                connectorId.getCatalogName(),
+                sessionPropertyManager,
+                warningCollector);
+    }
```
warning collector will be "final" in session; so you don't need this.

feilong-liu replied:
Deleted as we include warningCollector in session now.

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/ConnectorSession.java:
```diff
@@ -58,4 +58,9 @@ default boolean isReadConstraints()
         return false;
     }
+
+    default void addWarningMessage(PrestoWarning warning)
+    {
+        return;
+    }
```
Just `getWarningCollector`. This will not be a default interface as the collector is always available.

feilong-liu replied:
Change to getWarningCollector to return warningCollector.

highker reviewed (commented):
Can you add a unit test to make sure we do get the warning in the end?

Extra (that can be a separate PR):
Given the collector is in the session now, can you also add a follow up to clean up many other tunneling in our codebase where both session and warning collector are passed through the interfaces? One example is `SplitSchedulingContext` for `HiveSplitManager`. We don't need the warning collector to be inside `SplitSchedulingContext` because session is available in `HiveSplitManager`.

feilong-liu commented on presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java:
```diff
@@ -376,7 +379,8 @@ public PlanNode visitFilter(FilterNode filter, Void context)
             // replaceExpression() may further optimize the expression; if the resulting expression is always false, then return empty Values node
             if (FALSE_CONSTANT.equals(replacedExpression)) {
-                return new ValuesNode(tableScan.getSourceLocation(), idAllocator.getNextId(), tableScan.getOutputVariables(), ImmutableList.of());
+                session.addWarningMessage(new PrestoWarning(HIVE_TABLESCAN_CONVERTED_TO_VALUESNODE, format("Table '%s' returns 0 rows, and is converted to an empty ValuesNode by HiveFilterPushdown", tableScan.getTable().toString())));
```
Done

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java:
```diff
@@ -376,7 +376,7 @@ public PlanNode visitFilter(FilterNode filter, Void context)
             // replaceExpression() may further optimize the expression; if the resulting expression is always false, then return empty Values node
             if (FALSE_CONSTANT.equals(replacedExpression)) {
-                return new ValuesNode(tableScan.getSourceLocation(), idAllocator.getNextId(), tableScan.getOutputVariables(), ImmutableList.of());
+                return new ValuesNode(tableScan.getSourceLocation(), idAllocator.getNextId(), tableScan.getOutputVariables(), ImmutableList.of(), Optional.of(tableScan.getTable().toString()));
```
`getTable()` -> `getTable().getConnectorHandle()`. Same for other places in the file

feilong-liu replied:
Done

highker commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java:
```diff
@@ -1108,6 +1117,13 @@ public void testFilterFunctionsWithOptimization()
+    @Test
+    public void testOptimizeWithWarningMessage()
+    {
+        assertWarnings(getQueryRunner(), getSession(), "SELECT * FROM lineitem WHERE orderkey = 1 and orderkey = 2",
+                ImmutableSet.of(HIVE_TABLESCAN_CONVERTED_TO_VALUESNODE.toWarningCode()));
```
nit: one param per inline with the first line empty:
```java
assertWarnings(
        getQueryRunner(),
        getSession(),
        "SELECT * FROM lineitem WHERE orderkey = 1 and orderkey = 2",
        ImmutableSet.of(HIVE_TABLESCAN_CONVERTED_TO_VALUESNODE.toWarningCode()));

feilong-liu replied:
Done

highker commented on presto-main/src/main/java/com/facebook/presto/server/QuerySessionSupplier.java:
```diff
@@ -115,6 +119,11 @@ else if (context.getTimeZoneId() != null) {
         }
 
+        // Put after setSystemProperty are called
+        WarningCollector warningCollector = warningCollectorFactory.create(
+                sessionBuilder.getSystemProperty(WARNING_HANDLING, WarningHandlingLevel.class));
```
We can put these two into the same line. Our line cap is at around 180 chars

feilong-liu replied:
Done

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/ConnectorSession.java:
```diff
@@ -58,4 +58,9 @@ default boolean isReadConstraints()
         return false;
     }
+
+    default WarningCollector getWarningCollector()
+    {
+        return WarningCollector.NOOP;
+    }
```
Let's remove default and implement the interface for all its implementations. "Default" is usually discouraged especially for such cases where all implementations have a clear return.

feilong-liu commented:
> Can you add a unit test to make sure we do get the warning in the end?

Added a unit test in presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java for it.

> Extra (that can be a separate PR): Given the collector is in the session now, can you also add a follow up to clean up many other tunneling in our codebase where both session and warning collector are passed through the interfaces? One example is `SplitSchedulingContext` for `HiveSplitManager`. We don't need the warning collector to be inside `SplitSchedulingContext` because session is available in `HiveSplitManager`.

Sure, will do in a separate pull request.

highker reviewed (approved):
LGTM % nits
